### ISSUE TITLE
Matching with \or patterns is not right

### ISSUE THREAD

**Issue by ttuegel** [kframework/kore#1245]
For example,

```
syntax ABC ::= A | B | C

syntax ABC ::= toC( ABC ) [function]
rule toC(A #Or B) => C
```

This rule is intended to say,

```
rule toC(A) => C
rule toC(B) => C
```

but it actually says `\equals(toC(A) ∨ toC(B), C)`, which could be satisfied if `toC(A) = ⊥` or `toC(B) = ⊥`. That is not what the user wanted!

See also: #699

**Comment by virgil-serbanuta** [comment]
So, the problem is that the axiom is not translated properly, right?

**Comment by ttuegel** [comment]
> So, the problem is that the axiom is not translated properly, right?

Yes. I don't know if we want to change the translation of functions in general, or just split the rule in the frontend, but the existing translation doesn't mean what the user wants it to mean.

**Comment by ttuegel** [comment]
To elaborate, for a rule (neglecting `requires` and `ensures`)

```
rule f(φ₁ ∨ φ₂) => ψ
```

the translation is

```
axiom f(φ₁ ∨ φ₂) = ψ
// equivalently,
axiom ⌊(f(φ₁) ∨ f(φ₂)) → ψ⌋ ∧ ⌊ψ → f(φ₁) ∨ f(φ₂)⌋
```

but the user meant

```
axiom ... ∧ ⌊ψ → f(φ₁) ∧ f(φ₂)⌋
```

These axioms are equivalent when all the patterns are functional; I think this is why there is a breakdown between the user's intuition and the logic.

**Comment by yzhang90** [comment]
```
rule toC(A #Or B) => C
```
gives less information than 
```
rule toC(A) => C
rule toC(B) => C
```
as previous definition doesn't say toC is defined on A.
I would say we should not allow user to write `#Or` in function definition rules or do some transformation in the frontend...

**Comment by ttuegel** [comment]
@dwightguth @traiansf How do you feel about disallowing `#Or` on the left-hand side of function rules?

**Comment by dwightguth** [comment]
If the conclusion is that we don't want to try to come up with a translation to kore for or patterns in function rules, it would be much better to desugar everything in the frontend than to simply disallow it. But I'm not opposed to that as a solution. That's essentially what the llvm backend does anyway, just at a slightly later phase.

**Comment by traiansf** [comment]
My only concern is that we shouldn't allow equality rules like the one above to reach KORE, as once it is in KORE we cannot "desugar" it into two rules.

**Comment by xc93** [comment]
I'm overall not in favor of allowing users to write `rule toC(A #Or B) => C` when they actually mean two rules `rule toC(A) => C` and `rule toC(B) => C`. The reason, as pointed out by @ttuegel and @yzhang90  , is that it is not logically correct unless `C` is a total function. 
Given that said, if we do have such `#Or` rules in the frontend, then the best way to translate them into backend would be one of the following:
1. Translate them into two separate rules; OR
2. Translate them into something like `axiom \exists Ch . toC( A /\ Ch=1 \/ B /\ Ch=2) = C`, where `Ch`, meaning "choice", has sort `Nat`.

**Comment by ttuegel** [comment]
> Translate them into something like `axiom \exists Ch . toC( A /\ Ch=1 \/ B /\ Ch=2) = C`, where `Ch`, meaning "choice", has sort `Nat`.

That's interesting. Can you explain how that helps? I'm not really in favor of this approach; we don't implement enough matching to do this, and I'm sure the LLVM backend doesn't either, but I am intrigued.

**Comment by xc93** [comment]
Suppose the goal is to compress two rules `toC(A) = C` and `toC(B) = C` into one. 

The problem with `toC(A \/ B) = C`, or by distribution law, `toC(A) \/ toC(B) = C`, where `_\/_` is the normal logical disjunction, is that `toC(A)` and `toC(B)` may overlap. In other words, the union between `toC(A)` and `toC(B)` is not disjoint, so it is not clear who "contributes" towards becoming `C`. 

To solve this problem, we need to make the direct union into a _disjoint_ union. And this is where the choice indicator `Ch` comes to play a role. By writing, `toC( A /\ Ch=1 \/ B /\ Ch=2)`, or by distribution law, `toC(A) /\ Ch = 1 \/ toC(B) /\ Ch = 2`, we manually separate the contribution from `toC(A)` and `toC(B)`, by assigning `Ch` to be different flags. And finally, we use `\exists` to quantify all choices, which yield the correct semantics.

In other words, the effect of the axiom (fixed a typo and modified to be precise):
```\forall Ch . Ch = 1 \/ Ch = 2 -> (toC( A /\ Ch=1 \/ B /\ Ch=2) = C)```
can be enumerated by instantiating `Ch` to be `1`, `2`, or any other numbers:
``` 
toC(A) = C // when Ch = 1
toC(B) = C // when Ch = 2
\bottom -> \whatever // owise
```

The trick can also be applied to a broder cases with any kinds of predicates about any finitely many patterns:
```\forall Ch . Ch = 1 \/ ... \/ Ch = 100 -> /\i ( Ch = i /\ Predicate(i) )```

**Comment by traiansf** [comment]
@xc93 nice trick!  I think something similar was used in the Scala prover to avoid splitting the configurations when doing symbolic execution.

**Comment by grosu** [comment]
@xc93 @traiansf @tomtomjhj yes, we planned to use this trick to replace the in-backend indexing implementation with a mathematically clean, probably correct theory transformation.  You can, in fact, replace all the rules with just one huge rule, but one which does matching optimally, without re-doing any work across rules.  For example, the two rules for `ite` become:
```
rule \exists Ch . its(true /\ Ch=1 \/ false /\ Ch=2, S1, S2) => S1 /\ Ch=1 \/ S2 /\ Ch=2.
```
This is still in the cards, but we have more urgent things to take care of.  It will resurface when we'll generate proof objects,

**Comment by grosu** [comment]
@tomtomjhj regarding your original question (which is what brought me here, thanks for pointing it out on slack), I may not be the right person to have an opinion about that.  After spending so many months working on matching logic, it is impossible for me to see any connection between 
```
rule toC(A #Or B) => C
```
and
```
rule toC(A) => C
rule toC(B) => C
```
The first thing that strikes is that `A` can itself be  `⊥` while `B` and `C` not; then the first theory is still consistent, while the second is not.  `toC` is not even needed to illustrate the difference.

Probably giving a warning is good regardless.

**Comment by dwightguth** [comment]
For what it's worth, I did not realize the difference logically between `A #Or B` and two rules when we implemented this feature. Given that fact, I am not opposed to changing the syntax of or-patterns in K so that it's different from the ML disjunction, but we do still need to support or patterns.

**Comment by ttuegel** [comment]
> I am not opposed to changing the syntax of or-patterns in K so that it's different from the ML disjunction

I think this solution keeps everyone happy.

**Comment by ttuegel** [comment]
@xc93 Can we think of a way to do the axiom translation locally? I mean, we have this, which is in some sense a global transformation (at the axiom's top-level):

```
\forall Ch . Ch = 1 \/ Ch = 2 -> (toC( A /\ Ch=1 \/ B /\ Ch=2) = C)
```

but we would really prefer to write something like this,

```
toC( \disjoint_union(A, B) ) = C
```

where `\disjoint_union` is some alias. This would, I think, minimize the changes necessary in the LLVM backend because we would just be swapping `\or` out for another token. Is that even possible?

**Comment by dwightguth** [comment]
In terms of the llvm backend, either desugaring everything in the frontend or else replacing \or with simply another symbol would be the two preferred approaches.

**Comment by xc93** [comment]
> @xc93 Can we think of a way to do the axiom translation locally? I mean, we have this, which is in some sense a global transformation (at the axiom's top-level):
> 
> ```
> \forall Ch . Ch = 1 \/ Ch = 2 -> (toC( A /\ Ch=1 \/ B /\ Ch=2) = C)
> ```
> 
> but we would really prefer to write something like this,
> 
> ```
> toC( \disjoint_union(A, B) ) = C
> ```
> 
> where `\disjoint_union` is some alias. This would, I think, minimize the changes necessary in the LLVM backend because we would just be swapping `\or` out for another token. Is that even possible?

I cannot think of a short solution to that for now.

**Comment by ttuegel** [comment]
@dwightguth and I agreed on a solution to this problem: function rules can be translated as follows:

```
rule fun(φ₁, φ₂, ...) => ψ requires Pre ensures Post
```

becomes:

```
axiom \and(Pre = true, X₁ ∈ φ₁ ∧ X₂ ∈ φ₂ ∧ ...)) → \and(fun(X₁, X₂, ...) = ψ, Post = true)
```

This would also fix #1146 and #1630.

To make this change in the backend, we need to recognize the conjunction of `\in` predicates in the pre-condition and evaluate it before instantiating the equation.
